protobuf/grpc_service.proto

```diff
@@ -1013,6 +1013,45 @@ message InferStatistics
   StatisticDuration cache_miss = 8;
 }
 
+//@@
+//@@.. cpp:var:: message InferResponseStatistics
+//@@
+//@@   Statistics per decoupled response.
+//@@
+message InferResponseStatistics
+{
+  //@@  .. cpp:var:: StatisticDuration compute_infer
+  //@@
+  //@@     The count and cumulative duration to compute a response.
+  //@@
+  StatisticDuration compute_infer = 1;
+
+  //@@  .. cpp:var:: StatisticDuration compute_output
+  //@@
+  //@@     The count and cumulative duration to extract the output tensors of a
+  //@@     response.
+  //@@
+  StatisticDuration compute_output = 2;
+
+  //@@  .. cpp:var:: StatisticDuration success
+  //@@
+  //@@     The count and cumulative duration for successful responses.
+  //@@
+  StatisticDuration success = 3;
+
+  //@@  .. cpp:var:: StatisticDuration fail
+  //@@
+  //@@     The count and cumulative duration for failed responses.
+  //@@
+  StatisticDuration fail = 4;
+
+  //@@  .. cpp:var:: StatisticDuration empty_response
+  //@@
+  //@@     The count and cumulative duration for empty responses.
+  //@@
+  StatisticDuration empty_response = 5;
```

Comment:
I was initially confused by the meaning of `flags_only`, because the entire context here is talking about `response` and it is not immediately clear what `flags` has to do with `response` until seeing the `TRITONBACKEND_ResponseSend()` API. Then, the user will have to infer from the API to see the only two things other than `flags` are `response` and `error`, so `flags_only` means `no response` and `no error`.

I am seeing another small corner case here is what if there is `no response` but an `error`. The current implementation will group it into `fail` regardless whether `compute_output_start = 0` or not.

Just to make the naming clearer, how about renaming it to `success_no_output` (or just `no_output`)?

cc @rmccorm4

Reply:
I think we landed on the term "empty" over "flags_only" when implementing the "empty_final_response" support. Lots of discussion here on it: https://github.com/triton-inference-server/server/pull/5878#issuecomment-1570486585.

So for consistency I think `ResponseFactorySend(nullptr /* response */, flags=0)` is "empty", and `ResponseFactorySend(nullptr /* response */, flags=TRITONSERVER_RESPONSE_COMPLETE_FINAL)` is "empty final".

For "error responses", I think we do actually send a response object: https://github.com/triton-inference-server/backend/blob/a06e9a1157d6b5b9b34b6d05a07bb84d517f17c9/src/backend_common.cc#L621-L623.

The lack of sending anything at all (not even calling the API to send a nullptr response or flags) I think is nothing and has no corresponding stat, though we can't stop the backend developer from calling the API anyway if they so choose.